universal-application-tool-0.0.1/app/services/program/ProgramServiceImpl.java

```diff
@@ -369,6 +372,36 @@ public ProgramDefinition deleteBlock(long programId, long blockDefinitionId)
     return programMaybe.get().getApplications();
   }
 
+  @Override
+  public void publishProgram(long id) throws ProgramNotFoundException {
+    try {
+      this.publishProgramAsync(id).toCompletableFuture().join();
+    } catch (CompletionException e) {
+      if (e.getCause() instanceof ProgramNotFoundException) {
+        throw new ProgramNotFoundException(id);
+      }
+      throw e;
+    }
+  }
+
+  @Override
+  public CompletionStage<Void> publishProgramAsync(long id) {
+    return programRepository
+        .lookupProgram(id)
+        .thenApplyAsync(
+            programMaybe ->
+                programMaybe.orElseThrow(
+                    () -> new RuntimeException(new ProgramNotFoundException(id))))
+        .thenComposeAsync(programRepository::publishProgramAsync);
+  }
+
+  @Override
+  public ProgramDefinition newDraftFrom(long id) throws ProgramNotFoundException {
+    ProgramDefinition program =
+        this.getProgramDefinition(id).toBuilder().setLifecycleStage(LifecycleStage.DRAFT).build();
```

Comment:
Question (blocking): This copies all of the question references from the previous program, which won't be in a DRAFT state. How will the draft program get editable questions?

Reply:
> we can create draft questions when questions are edited
so copy-on-write?

Reply:
Yep!